File: src/db/codegen.js
#!/usr/bin/env node
import { db } from './db'; // Ensure this points to your database connection file
import { inviteCodesTable } from './schema'; // Ensure this points to your schema file
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';

const argv = yargs(hideBin(process.argv)).argv;

function generateInviteCode(length = 10) {
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    const charactersLength = characters.length;
    for (let i = 0; i < length; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
}

async function populateInviteCodes(tier, count) {
    const codes = [];
    for (let i = 0; i < count; i++) {
        codes.push({
            code: generateInviteCode(),
            tier: tier
        });
    }

    for (const code of codes) {
        await db.insertInto(inviteCodesTable).values(code).execute();
    }

    console.log(`${count} ${tier} invite codes have been populated`);
}

async function main() {
    const tiers = ['free', 'bronze', 'silver', 'gold', 'premium', 'master', 'elite'];
    
    for (const tier of tiers) {
        if (argv[tier]) {
            const count = parseInt(argv[tier], 10);
            await populateInviteCodes(tier, count);
        }
    }
}

main().catch(console.error);

File: src/db/codegen.py
import psycopg2
from psycopg2 import sql
import random
import string
import argparse

def generate_invite_code(length=10):
    characters = string.ascii_letters + string.digits
    return ''.join(random.choice(characters) for _ in range(length))

def populate_invite_codes(conn, tier, count):
    codes = [(generate_invite_code(), tier) for _ in range(count)]

    with conn.cursor() as cur:
        insert_query = sql.SQL("INSERT INTO invite_codes (code, tier) VALUES (%s, %s)")
        cur.executemany(insert_query, codes)
        conn.commit()

    print(f"{count} {tier} invite codes have been populated")

def main(connection_uri, free, bronze, silver, gold, premium, master, elite):
    conn = psycopg2.connect(connection_uri)

    try:
        tiers = {
            'free': free,
            'bronze': bronze,
            'silver': silver,
            'gold': gold,
            'premium': premium,
            'master': master,
            'elite': elite
        }

        for tier, count in tiers.items():
            if count > 0:
                populate_invite_codes(conn, tier, count)
    finally:
        conn.close()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Populate invite codes into the database")
    parser.add_argument('--connection-uri', required=True, help='PostgreSQL connection URI')
    parser.add_argument('--free', type=int, default=0, help='Number of free invite codes to generate')
    parser.add_argument('--bronze', type=int, default=0, help='Number of bronze invite codes to generate')
    parser.add_argument('--silver', type=int, default=0, help='Number of silver invite codes to generate')
    parser.add_argument('--gold', type=int, default=0, help='Number of gold invite codes to generate')
    parser.add_argument('--premium', type=int, default=0, help='Number of premium invite codes to generate')
    parser.add_argument('--master', type=int, default=0, help='Number of master invite codes to generate')
    parser.add_argument('--elite', type=int, default=0, help='Number of elite invite codes to generate')

    args = parser.parse_args()
    main(args.connection_uri, args.free, args.bronze, args.silver, args.gold, args.premium, args.master, args.elite)
